### DISCUSSION TITLE
Gradle run error and failure to pass `skip checkup`

### DISCUSSION THREAD

**Discussion by Harryuxd** [skiptools/skip#4] [Q&A] [answered]
This error came up when trying to build in xcode
I followed all the steps and made 2 different projects but same error in xcode
I ran skip doublecheck here
<img width="900" alt="Screenshot 2023-10-26 at 11 18 07 PM" src="https://github.com/skiptools/skip/assets/47468239/cedf8f33-138c-40fb-9004-be57e3272074">

**Comment by joshbirnholz** [comment]
Same issue here!

<img width="1146" alt="Screenshot 2023-10-27 at 1 16 40 AM" src="https://github.com/skiptools/skip/assets/4167834/1d44863e-dce6-48d5-9023-a03bce84d201">

**Comment by Marsgames** [comment]
Same here
![image](https://github.com/skiptools/skip/assets/9108765/eaa83e43-2f4a-440e-8069-0b5830193624)

**Comment by johankool** [comment]
Similar issue here. When I run ` skip checkup -vEM ` I get more info.

```
GRADLE> 
GRADLE> FAILURE: Build failed with an exception.
GRADLE> 
GRADLE> * What went wrong:
GRADLE> Could not determine the dependencies of task ':HelloSkip:testDebugUnitTest'.
GRADLE> > Could not create task ':HelloSkip:compileDebugUnitTestJavaWithJavac'.
GRADLE>    > Failed to calculate the value of task ':HelloSkip:compileDebugUnitTestJavaWithJavac' property 'javaCompiler'.
GRADLE>       > No matching toolchains found for requested specification: {languageVersion=17, vendor=any, implementation=vendor-specific} for MAC_OS on aarch64.
GRADLE>          > No locally installed toolchains match and toolchain download repositories have not been configured.
GRADLE> 
GRADLE> * Try:
GRADLE> > Learn more about toolchain auto-detection at https://docs.gradle.org/8.4/userguide/toolchains.html#sec:auto_detection.
GRADLE> > Learn more about toolchain repositories at https://docs.gradle.org/8.4/userguide/toolchains.html#sub:download_repositories.
GRADLE> > Run with --stacktrace option to get the stack trace.
GRADLE> > Run with --info or --debug option to get more log output.
GRADLE> > Run with --scan to get full insights.
GRADLE> > Get more help at https://help.gradle.org.
```

If I go to the `... /plugins/outputs/hello-skip/HelloSkip/skipstone` folder and run `gradle -q javaToolchains` I see:
```
 + Options
     | Auto-detection:     Enabled
     | Auto-download:      Enabled

 + Homebrew JDK 21.0.1
     | Location:           /opt/homebrew/Cellar/openjdk/21.0.1/libexec/openjdk.jdk/Contents/Home
     | Language Version:   21
     | Vendor:             Homebrew
     | Architecture:       aarch64
     | Is JDK:             true
     | Detected by:        Current JVM
```

I tried to install version 17 as that seems to be required with `brew install openjdk@17 ` but I can't seem to get gradle to notice that toolchain.

**Reply by johankool** [reply]
I got it to work by setting adding JAVA_HOME build setting in the Xcode project with the value for the OpenJDK 17 home: `/opt/homebrew/opt/openjdk@17/libexec/openjdk.jdk/Contents/Home`.

**Reply by Harryuxd** [reply]
Hi what steps do i need to take to do this?

**Reply by davidakoontz** [reply]
@johankool Johankool  what is the magic file name that you set JAVA_HOME for Skip/Gradle error?  Inquiring Minds want to do the same!

**Reply by johankool** [reply]
@oluarry and @davidakoontz Select your project in Xcode, go to the target build settings, add a user-defined setting, name it JAVA_HOME and set the value to `/opt/homebrew/opt/openjdk@17/libexec/openjdk.jdk/Contents/Home` (on Apple Silicon Macs, when installed with `brew install openjdk@17`).

**Reply by davidakoontz** [reply]
After a two days of chasing this problem... I've got to YELL into the Void....  I HATE JAVA!

I have the same errors when I go back to the original Hello-Skip project that worked some days ago...

I finally came to think that all my attempts to SET the proper JAVA / ANDROID / JDK / SDK etc.  was causing as much PAIN as not setting them...  I remembered that Yes - I'd had a Mac before this M1 MacBook and it had left-over bits of Java moved over when I migrated systems from old Intel MacBook - to new M1 MacBook... and cleaning that shit right out was going to lessen the smell of JAVA.  So I did that - deleted some Java directories... leaving only ONE. 
Then I set about UNDOING all the ENV VAR that I had set.  Cleaning out the .zprofile & forgetting some... I'm sure.

Then I closed all my terminals (because they reload the profiles at launch) and tried Skip in the original Hello-Skip project that worked on my machine  AGAIN, Now! Better than before... so maybe I'm onto something here...

And yes even the new Hello-Skip with slight folder names for my desire project... started compiling no-errors!

I think riding my machine of old Intel install Java instances was key to resolving my problems - and setting even-more Env Vars - with no way to validate they are correct is a BAD Solution.

**Reply by davidakoontz** [reply]
Thank you Johan!  I came at it form a different direction... cleaned my machine of old (Intel) Java Installs and poof <magic> it started to work...

It is a mystery - when Java works & a bigger mystery when it fails.  I've got a new view of the lack of validation that setting ENV VARS to "fix" a problem is fraught with thorns and a fragile system...  UNIX / Linux  when you have a few... OK / When you have hundreds... not a solutions!

**Reply by marcprux** [reply]
Thanks for sharing your insights, and welcome to the wild world of Java development. We hope the experience improves for you.

I suspect that what happened was that an old Intel version of Java was being picked up, possibly due to an old install of Homebrew. Homebrew needed to move from `/usr/local/homebrew/` on Intel hosts to `/opt/homebrew/` on ARM hosts. The upshot of this is that a lot of users who migrated at some point in the past from an Intel mac to an ARM mac may still have vestigial PATH and other environment references to previous Intel-only installations of some tools. This has affected a number of Homebrew users since the M1 macs started shipping.

The good news is that this problem will be mitigated somewhat when Gradle 8.5 is released (hopefully next month), which will enable the Skip Homebrew package to depend on OpenJDK 21, which makes it less likely that a stale JVM (with a non-native architecture) will be found in the path. We will also endeavor to try to identify these sorts of issues with the `skip doctor` and `skip checkup` commands, and provide some guidance to the user when these circumstances are detected.

**Comment by Marsgames** [comment, marked as answer]
Solution found in #5
